File: Heaps/maxHeap/Java/MaxHeap.java
import java.util.NoSuchElementException;

/**
 * Implementation of a Max-Heap.
 * A Max-Heap is a specialization of a Heap that is used for
 * the implementation of many algorithms, especially on Graphs.
 * The difference to a normal Heap is that the root is always
 * the global maximum of all elements.
 * To achieve that a "bubble-up" will be performed if necessary.
 *
 * @author Thomas Lang
 * @version July 15, 2014
 * @see <a href="http://en.wikipedia.org/wiki/Min-max_heap">Min-max heap</a>
 */
public class MaxHeap<T extends Comparable<T>> extends Heap<T>{

    /**
     * Constructor for a new Max-Heap, 
     * just calls the Heap-constructor.
     */
    public MaxHeap(){
        super();
    }

    /**
     * Inserts a new Node containing <code>value</code> to
     * the Heap. The newly created Node will be inserted via
     * the <code>insert</code>-method in the abstract class 
     * and then a bubble-up is performed.
     *
     * @param  value  The value of the new Node to insert.
     */
    public void insert( T value ){
        Node<T> node = new Node<T>( value );
        super.insert( node );
        bubbleUp( node );
    }

    /**
     * Performs the "delete max" operation on the Heap.
     * This means that the root (that is always the maximum
     * of all elements) will be swapped with the last node
     * inserted. After that, that node will be deleted and
     * the new root will be "bubble-down'ed" until 
     * everything is okay again.
     *
     * @throws NoSuchElementException
     */
    public void deleteMax() throws NoSuchElementException{
        Node<T> last = getLastNode();
        root.setValue( last.getValue() );
        // deleting "last" node
        if( last.getParent().getLeft() == last ){
            last.getParent().setLeft( null );
            last.setParent( null );
        }else{
            last.getParent().setRight( null );
            last.setParent( null );
        }
        size--;

        bubbleDown( root );
    }

 
    /**
     * Does the "bubble down".
     * So it swaps a passed <code>node</code> with either
     * its left or right child until the ordering in the
     * Heap is restored.
     *
     * @param  node  Current node.
     */
    private void bubbleDown( Node<T> node ){
        // if node has no children, we cannot make bubbleDown
        if(( node.getLeft() == null ) && ( node.getRight() == null ))
            return;

        // if node has left and right children
        if(( node.getLeft() != null ) && ( node.getRight() != null )){
            if(( node.getValue().compareTo( node.getLeft().getValue() ) < 0 ) &&
               ( node.getLeft().getValue().compareTo( node.getRight().getValue() ) > 0 ))
            {
                swapNodes( node, node.getLeft() );
                bubbleDown( node.getLeft() );
            }else if( node.getValue().compareTo( node.getRight().getValue() ) < 0 ){
                swapNodes( node, node.getRight() );
                bubbleDown( node.getRight() );
            }
        }

        // if left child only
        if(( node.getLeft() != null ) && ( node.getValue().compareTo( node.getLeft().getValue() ) < 0 )){
                swapNodes( node, node.getLeft() );
                bubbleDown( node.getLeft() );
        }

        // if right child only
        if(( node.getRight() != null ) && ( node.getValue().compareTo( node.getRight().getValue() ) < 0 )){
                swapNodes( node, node.getRight() );
                bubbleDown( node.getRight() );
            }
    }


    /**
     * Performs the "bubble-Up" operation. So it checks, if
     * the parent Node is smaller. If yes, the two will be
     * swapped and then this goes recursively through the
     * entire Heap.
     * So we can ensure, that the root of the Heap is always
     * the global maximum of all elements.
     *
     * @param  bNode  The Node that will be "bubble-up'ed".
     */
    private void bubbleUp( Node<T> bNode ){
        if( bNode.getParent() == null )
            return;
        if( bNode.getValue().compareTo( bNode.getParent().getValue() ) > 0 ){
            swapNodes( bNode, bNode.getParent() );
            bubbleUp( bNode.getParent() );
        }
    }

    /**
     * Helper method for swapping two Nodes.
     *
     * @param  child   The first Node to swap.
     * @param  parent  The second Node to swap.
     */
    private void swapNodes( Node<T> child, Node<T> parent ){
        T childValue = child.getValue();
        child.setValue( parent.getValue() );
        parent.setValue( childValue );
    }
}

File: Heaps/maxHeap/Python/MaxHeap.py

from Heap import *

class MaxHeap( Heap ):
    """Specialized form of the abstract Heap.
       This one follows the special rule, that
       a Node is always greater (by value) than
       its children, there is still no connection
       between its children. This has the 
       advantage that the root of this tree is
       ALWAYS THE GLOBAL MAXIMUM of the entire Heap."""

    def __init__( self ):
        """Creates a new but empty MaxHeap."""
        super( MaxHeap, self ).__init__()


    def insert( self, value ):
        """Inserts a new Node containing the passed
           value to the tree and then performs a
           bubbleUp (if necessary).
           
        @param  value  The value to add.    """
        node = Node( value )
        super( MaxHeap, self ).insert( node )
        self.__bubbleUp( node )


    def __bubbleUp( self, node ):
        """Performs a 'bubbleUp' operation.
           That means that it swaps a node 
           with its parent Node if the 
           condition mentioned above is
           violated.
           
        @param  node  The node to bubbleUp"""
        if node.parent is None:
            return
        if node.value > node.parent.value:
            self.__swapNodes( node, node.parent )
            self.__bubbleUp( node.parent )


    def __swapNodes( self, child, parent ):
        """Helper method, swaps the two
           passed Nodes by value only.
           
        @param  child  The child Node.
        @param  parent The parent Node. """
        childValue = child.value
        child.setValue( parent.value )
        parent.setValue( childValue )


    def deleteMax( self ):
        """Deletes the maximum Node that is
           the root of the Heap (see 
           class docstring ).          """
        last = self.getLastNode()
        self.root.setValue( last.value )

        if last.parent.left == last:
            last.parent.setLeft( None )
            last.setParent( None )
        else:
            last.parent.setRight( None )
            last.setParent( None )

        self.size -= 1
        self.__bubbleDown( self.root )


    def __bubbleDown( self, node ):
        """Performs a 'bubbleDown' operation, so it
           corrects any misstandings in the Heap
           caused by the 'deleteMax' operation. 
           
        @param  node  The node to bubbleDown.    """
        if node.left is None and node.right is None:
            return

        if node.left is not None and node.right is not None:
            if node.value < node.left.value:
                self.__swapNodes( node, node.left )
                self.__bubbleDown( node.left )
            elif node.value < node.right.value:
                self.__swapNodes( node, node.right )
                self.__bubbleDown( node.right )

        if node.left is not None and node.value < node.left.value:
            self.__swapNodes( node, node.left )
            self.__bubbleDown( node.left )

        if node.right is not None and node.value < node.right.value:
            self.__swapNodes( node, node.right )
            self.__bubbleDown( node.right )

